### ISSUE TITLE
torch.nn.functional.scaled_dot_product_attention() : support both attn_mask and is_causal

### ISSUE THREAD

**Issue by pfeatherstone** [pytorch/pytorch#101531]
### 🚀 The feature, motivation and pitch

It would still be great if `torch.nn.functional.scaled_dot_product_attention()` supported setting both `attn_mask` and `is_causal=True`. In which case it ignores the upper triangular part of `attn_mask` and implicitly assumes it's either set to False or -inf. This would allow us to do some sort of relative positional encoding and causal masking at the same time while saving compute and memory.

### Alternatives

At the moment you set `is_causal=False` and make sure `attn_mask` is inherently causal. That's fine, but we're missing an optimization.

### Additional context

Mentioned in #100709

**Comment by vadimkantorov** [comment]
Curious, is general attn_bias of shape `(batch_size, num_heads, seq_len, seq_len)` or broadcasted `(batch_size, 1, seq_len, seq_len)` or `(1, 1, seq_len, seq_len)` supported now by SDPA?

If so, Nvidia Nemo Conformer could finally transition to using SDPA: 
- https://github.com/NVIDIA/NeMo/issues/6257
